File: app/src/main/java/com/sqube/tipshub/MyProfileActivity.java
package com.sqube.tipshub;

import android.app.AlertDialog;
import android.content.Intent;
import android.os.Bundle;
import com.google.android.material.tabs.TabLayout;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentPagerAdapter;
import androidx.viewpager.widget.ViewPager;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.bumptech.glide.Glide;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.firestore.DocumentSnapshot;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;

import adapters.PerformanceAdapter;
import de.hdodenhof.circleimageview.CircleImageView;
import fragments.BankersFragment;
import fragments.PostFragment;
import fragments.ReviewFragment;
import models.ProfileMedium;
import utils.FirebaseUtil;
import utils.Reusable;

import static utils.Reusable.getPlaceholderImage;

public class MyProfileActivity extends AppCompatActivity implements View.OnClickListener {
    private ActionBar actionBar;
    private String userId, imgUrl;
    private CircleImageView imgDp;
    private LinearLayout[] lnrLayout = new LinearLayout[4];
    ProfileMedium profile;
    private RecyclerView recyclerView;
    Fragment postFragment, bankerFragment, reviewFragment;
    ArrayList<Map<String, Object>> performanceList = new ArrayList<>();
    private TextView txtName, txtUsername, txtBio;
    private TextView txtPost, txtWon, txtAccuracy;
    private TextView txtFollowers, txtFollowing, txtSubscribers, txtSubscriptions;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_my_profile);
        actionBar = getSupportActionBar();
        if (actionBar != null) {
            actionBar.setDisplayHomeAsUpEnabled(true);
        }
        ViewPager viewPager = findViewById(R.id.viewpager);
        TabLayout tabLayout = findViewById(R.id.tabs);
        tabLayout.setupWithViewPager(viewPager);
        imgDp = findViewById(R.id.imgDp);
        txtName = findViewById(R.id.txtFullName);
        txtUsername = findViewById(R.id.txtUsername);
        txtBio = findViewById(R.id.txtBio);
        txtPost = findViewById(R.id.txtPost);
        txtWon = findViewById(R.id.txtWon);
        txtAccuracy = findViewById(R.id.txtAccuracy);
        txtFollowers = findViewById(R.id.txtFollowers);
        txtFollowing = findViewById(R.id.txtFollowing);
        txtSubscribers = findViewById(R.id.txtSubscribers);
        txtSubscriptions = findViewById(R.id.txtSubscription);
        lnrLayout[0] = findViewById(R.id.lnrFollowing);
        lnrLayout[1] = findViewById(R.id.lnrFollowers);
        lnrLayout[2] = findViewById(R.id.lnrSubscribers);
        lnrLayout[3] = findViewById(R.id.lnrSubscription);
        for(LinearLayout l: lnrLayout)
            l.setOnClickListener(this);
        recyclerView = findViewById(R.id.performanceList);
        LinearLayoutManager lm = new LinearLayoutManager(getApplicationContext());
        TextView btnEdit = findViewById(R.id.btnEdit);
        btnEdit.setOnClickListener(v -> startActivity(new Intent(MyProfileActivity.this, SettingsActivity.class)));
        recyclerView.setLayoutManager(lm);
        FirebaseUser user = FirebaseUtil.getFirebaseAuthentication().getCurrentUser();
        userId = user.getUid();
        setupViewPager(viewPager); //set up view pager with fragments
    }

    private  Map<String, Object> getRow(int i) {
        Map<String, Object> row = new HashMap<>();
        switch (i){
            case 1:
                if(profile.getE1a_NOG()>0){
                    row.put("type", i);
                    row.put("NOG", profile.getE1a_NOG());
                    row.put("WG", profile.getE1b_WG());
                    row.put("WGP", profile.getE1c_WGP());
                }
                break;
            case 2:
                if(profile.getE2a_NOG()>0){
                    row.put("type", i);
                    row.put("NOG", profile.getE2a_NOG());
                    row.put("WG", profile.getE2b_WG());
                    row.put("WGP", profile.getE2c_WGP());
                }
                break;
            case 3:
                if(profile.getE3a_NOG()>0){
                    row.put("type", i);
                    row.put("NOG", profile.getE3a_NOG());
                    row.put("WG", profile.getE3b_WG());
                    row.put("WGP", profile.getE3c_WGP());
                }
                break;
            case 4:
                if(profile.getE4a_NOG()>0){
                    row.put("type", i);
                    row.put("NOG", profile.getE4a_NOG());
                    row.put("WG", profile.getE4b_WG());
                    row.put("WGP", profile.getE4c_WGP());
                }
                break;
            case 5:
                if(profile.getE5a_NOG()>0){
                    row.put("type", i);
                    row.put("NOG", profile.getE5a_NOG());
                    row.put("WG", profile.getE5b_WG());
                    row.put("WGP", profile.getE5c_WGP());
                }
                break;
            case 6:
                if(profile.getE6a_NOG()>0){
                    row.put("type", i);
                    row.put("NOG", profile.getE6a_NOG());
                    row.put("WG", profile.getE6b_WG());
                    row.put("WGP", profile.getE6c_WGP());
                }
                break;
        }
        return row;
    }

    private void setupViewPager(ViewPager viewPager) {
        Bundle bundle = new Bundle();
        bundle.putString("userId", userId);
        postFragment = new PostFragment();
        bankerFragment = new BankersFragment();
        reviewFragment = new ReviewFragment();
        //passing bunder with userId to fragment
        postFragment.setArguments(bundle);
        bankerFragment.setArguments(bundle);
        reviewFragment.setArguments(bundle);
        ViewPagerAdapter adapter = new ViewPagerAdapter(getSupportFragmentManager());
        adapter.addFragment(postFragment, "Posts");
        adapter.addFragment(bankerFragment, "Bankers");
        adapter.addFragment(reviewFragment, "Review");
        viewPager.setAdapter(adapter);
    }

    public class ViewPagerAdapter extends FragmentPagerAdapter{
        private final List<Fragment> mFragmentList = new ArrayList<>();
        private final List<String> mFragmentTitleList = new ArrayList<>();

        public ViewPagerAdapter(FragmentManager manager) {
            super(manager);
        }

        @Override
        public Fragment getItem(int position) {
            return mFragmentList.get(position);
        }

        @Override
        public int getCount() {
            return mFragmentList.size();
        }

        public void addFragment(Fragment fragment, String title) {
            mFragmentList.add(fragment);
            mFragmentTitleList.add(title);
        }

        @Override
        public CharSequence getPageTitle(int position) {
            return mFragmentTitleList.get(position);
        }
    }

    private void showDp() {
        AlertDialog.Builder builder = new AlertDialog.Builder(MyProfileActivity.this);
        LayoutInflater inflater = this.getLayoutInflater();
        View dialogView;
        dialogView = inflater.inflate(R.layout.image_viewer, null);
        builder.setView(dialogView);
        final AlertDialog dialog= builder.create();
        dialog.show();
        ImageView imgProfile = dialog.findViewById(R.id.imgDp);
        //set Display picture
        Glide.with(getApplicationContext()).load(profile.getB2_dpUrl())
                .placeholder(R.drawable.dummy)
                .error(getPlaceholderImage(userId.charAt(0)))
                .into(imgProfile);
    }
    @Override
    protected void onPostResume() {
        super.onPostResume();
        FirebaseUtil.getFirebaseFirestore().collection("profiles").document(userId).get()
                .addOnSuccessListener(new OnSuccessListener<DocumentSnapshot>() {
            @Override
            public void onSuccess(DocumentSnapshot documentSnapshot) {
                if(documentSnapshot==null||!documentSnapshot.exists())
                    return;
                profile = documentSnapshot.toObject(ProfileMedium.class);
                imgUrl = profile.getB2_dpUrl();
                if(!imgUrl.isEmpty())
                    imgDp.setOnClickListener(MyProfileActivity.this);
                String name = String.format(Locale.getDefault(),"%s %s", profile.getA0_firstName(),profile.getA1_lastName());
                actionBar.setTitle(name);
                txtName.setText(name);
                txtUsername.setText(String.format(Locale.getDefault(),"@%s",profile.getA2_username()));
                txtBio.setText(profile.getA5_bio());
                Reusable.applyLinkfy(MyProfileActivity.this, profile.getA5_bio(), txtBio);
                txtFollowers.setText(String.valueOf(profile.getC4_followers()));
                txtFollowing.setText(String.valueOf(profile.getC5_following()));
                txtSubscribers.setText(String.valueOf(profile.getC6_subscribers()));
                txtSubscriptions.setText(String.valueOf(profile.getC7_subscriptions()));
                String tips = profile.getE0a_NOG()>1? "tips": "tip";
                txtPost.setText(String.format(Locale.getDefault(),"%d  %s  • ", profile.getE0a_NOG(), tips));
                txtWon.setText(String.format(Locale.getDefault(),"%d  won  • ", profile.getE0b_WG()));
                txtAccuracy.setText(String.format(Locale.getDefault(),"%.1f%%", (double) profile.getE0c_WGP()));

                //set Display picture
                Glide.with(getApplicationContext()).load(profile.getB2_dpUrl())
                        .placeholder(R.drawable.dummy)
                        .error(getPlaceholderImage(userId.charAt(0)))
                        .into(imgDp);

                if(!performanceList.isEmpty())
                    return;
                if(profile.getE0a_NOG()>0){
                    for(int i=1; i<=6; i++){
                        Map<String, Object> row = getRow(i);
                        if(!row.isEmpty())
                            performanceList.add(row);
                    }
                    recyclerView.setAdapter(new PerformanceAdapter(performanceList));
                }
            }
        });
    }

    @Override
    public void onClick(View v) {
        Intent intent = new Intent(this, FollowerListActivity.class);
        intent.putExtra("personId", userId);
        switch (v.getId()){
            case R.id.imgDp:
                showDp();
                break;
            case R.id.lnrFollowers:
                if(Integer.valueOf(txtFollowers.getText().toString())<1)
                    return;
                intent.putExtra("search_type", "followers");
                startActivity(intent);
                break;
            case R.id.lnrFollowing:
                if(Integer.valueOf(txtFollowing.getText().toString())<1)
                    return;
                intent.putExtra("search_type", "followings");
                startActivity(intent);
                break;
            case R.id.lnrSubscribers:
                if(Integer.valueOf(txtSubscribers.getText().toString())<1)
                    return;
                intent.putExtra("search_type", "subscribers");
                startActivity(intent);
                break;
            case R.id.lnrSubscription:
                if(Integer.valueOf(txtSubscriptions.getText().toString())<1)
                    return;
                intent.putExtra("search_type", "subscribed_to");
                startActivity(intent);
                break;
        }

    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        finish();
        return true;
    }


    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
    }
}

File: app/src/main/java/com/sqube/tipshub/activities/MyProfileActivity.kt
package com.sqube.tipshub.activities

import com.sqube.tipshub.utils.FirebaseUtil.firebaseAuthentication
import com.sqube.tipshub.utils.Reusable.Companion.getPlaceholderImage
import com.sqube.tipshub.utils.FirebaseUtil.firebaseFirestore
import com.sqube.tipshub.utils.Reusable.Companion.applyLinkfy
import androidx.appcompat.app.AppCompatActivity
import com.sqube.tipshub.models.ProfileMedium
import android.widget.TextView
import android.os.Bundle
import androidx.viewpager.widget.ViewPager
import androidx.recyclerview.widget.LinearLayoutManager
import android.content.Intent
import com.sqube.tipshub.fragments.PostFragment
import com.sqube.tipshub.fragments.BankersFragment
import com.sqube.tipshub.fragments.ReviewFragment
import androidx.fragment.app.FragmentPagerAdapter
import android.graphics.drawable.ColorDrawable
import com.bumptech.glide.Glide
import com.google.firebase.firestore.DocumentSnapshot
import adapters.PerformanceAdapter
import android.app.AlertDialog
import android.graphics.Color
import android.view.MenuItem
import android.view.View
import android.widget.ImageView
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import com.sqube.tipshub.R
import com.sqube.tipshub.databinding.ActivityMyProfileBinding
import java.util.*

class MyProfileActivity : AppCompatActivity(), View.OnClickListener {
    private lateinit var _binding: ActivityMyProfileBinding
    private val binding get() = _binding
    private var userId: String? = null
    private var imgUrl: String? = null
    var profile: ProfileMedium? = null
    var performanceList = ArrayList<Map<String, Any>>()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        _binding = ActivityMyProfileBinding.inflate(layoutInflater)
        setContentView(binding.root)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        binding.tabs.setupWithViewPager(binding.viewpager)
        binding.lnrFollowing.setOnClickListener(this)
        binding.lnrFollowers.setOnClickListener(this)
        binding.lnrSubscribers.setOnClickListener(this)
        binding.lnrSubscription.setOnClickListener(this)
        val btnEdit = findViewById<TextView>(R.id.btnEdit)
        btnEdit.setOnClickListener { v: View? -> startActivity(Intent(this@MyProfileActivity, SettingsActivity::class.java)) }
        binding.performanceList.layoutManager = LinearLayoutManager(this)
        val user = firebaseAuthentication!!.currentUser
        userId = user!!.uid
        setupViewPager(binding.viewpager) //set up view pager with com.sqube.tipshub.fragments
    }

    private fun getRow(i: Int): Map<String, Any> {
        val row: MutableMap<String, Any> = HashMap()
        when (i) {
            1 -> if (profile!!.e1a_NOG > 0) {
                row["type"] = i
                row["NOG"] = profile!!.e1a_NOG
                row["WG"] = profile!!.e1b_WG
                row["WGP"] = profile!!.e1c_WGP
            }
            2 -> if (profile!!.e2a_NOG > 0) {
                row["type"] = i
                row["NOG"] = profile!!.e2a_NOG
                row["WG"] = profile!!.e2b_WG
                row["WGP"] = profile!!.e2c_WGP
            }
            3 -> if (profile!!.e3a_NOG > 0) {
                row["type"] = i
                row["NOG"] = profile!!.e3a_NOG
                row["WG"] = profile!!.e3b_WG
                row["WGP"] = profile!!.e3c_WGP
            }
            4 -> if (profile!!.e4a_NOG > 0) {
                row["type"] = i
                row["NOG"] = profile!!.e4a_NOG
                row["WG"] = profile!!.e4b_WG
                row["WGP"] = profile!!.e4c_WGP
            }
            5 -> if (profile!!.e5a_NOG > 0) {
                row["type"] = i
                row["NOG"] = profile!!.e5a_NOG
                row["WG"] = profile!!.e5b_WG
                row["WGP"] = profile!!.e5c_WGP
            }
            6 -> if (profile!!.e6a_NOG > 0) {
                row["type"] = i
                row["NOG"] = profile!!.e6a_NOG
                row["WG"] = profile!!.e6b_WG
                row["WGP"] = profile!!.e6c_WGP
            }
        }
        return row
    }

    private fun setupViewPager(viewPager: ViewPager) {
        val bundle = Bundle()
        bundle.putString("userId", userId)
        val postFragment = PostFragment()
        val bankerFragment = BankersFragment()
        val reviewFragment = ReviewFragment()
        //passing bunder with userId to fragment
        postFragment.arguments = bundle
        bankerFragment.arguments = bundle
        reviewFragment.arguments = bundle
        val adapter = ViewPagerAdapter(supportFragmentManager)
        adapter.addFragment(postFragment, "Posts")
        adapter.addFragment(bankerFragment, "Bankers")
        adapter.addFragment(reviewFragment, "Review")
        viewPager.adapter = adapter
    }

    inner class ViewPagerAdapter(manager: FragmentManager?) : FragmentPagerAdapter(manager!!) {
        private val mFragmentList: MutableList<Fragment> = ArrayList()
        private val mFragmentTitleList: MutableList<String> = ArrayList()
        override fun getItem(position: Int): Fragment {
            return mFragmentList[position]
        }

        override fun getCount(): Int {
            return mFragmentList.size
        }

        fun addFragment(fragment: Fragment, title: String) {
            mFragmentList.add(fragment)
            mFragmentTitleList.add(title)
        }

        override fun getPageTitle(position: Int): CharSequence? {
            return mFragmentTitleList[position]
        }
    }

    private fun showDp() {
        val builder = AlertDialog.Builder(this@MyProfileActivity)
        val inflater = this.layoutInflater
        val dialogView = inflater.inflate(R.layout.image_viewer, null)
        builder.setView(dialogView)
        val dialog = builder.create()
        Objects.requireNonNull(dialog.window)?.setBackgroundDrawable(ColorDrawable(Color.TRANSPARENT))
        dialog.show()
        val imgProfile = dialog.findViewById<ImageView>(R.id.imgDp)
        //set Display picture
        Glide.with(applicationContext).load(profile!!.b2_dpUrl)
                .placeholder(R.drawable.dummy)
                .error(getPlaceholderImage(userId!![0]))
                .into(imgProfile)
    }

    override fun onPostResume() {
        super.onPostResume()
        firebaseFirestore!!.collection("profiles").document(userId!!).get()
                .addOnSuccessListener { documentSnapshot: DocumentSnapshot? ->
                    if (documentSnapshot == null || !documentSnapshot.exists()) return@addOnSuccessListener
                    profile = documentSnapshot.toObject(ProfileMedium::class.java)
                    imgUrl = profile!!.b2_dpUrl
                    if (imgUrl!="") binding.imgDp.setOnClickListener(this@MyProfileActivity)
                    val name = String.format(Locale.getDefault(), "%s %s", profile!!.a0_firstName, profile!!.a1_lastName)
                    supportActionBar?.title = name
                    binding.txtFullName.text = name
                    binding.txtUsername.text = String.format(Locale.getDefault(), "@%s", profile!!.a2_username)
                    binding.txtBio.text = profile!!.a5_bio
                    applyLinkfy(this@MyProfileActivity, profile!!.a5_bio, binding.txtBio)
                    binding.txtFollowers.text = profile!!.c4_followers.toString()
                    binding.txtFollowing.text = profile!!.c5_following.toString()
                    binding.txtSubscribers.text = profile!!.c6_subscribers.toString()
                    binding.txtSubscription.text = profile!!.c7_subscriptions.toString()
                    val tips = if (profile!!.e0a_NOG > 1) "tips" else "tip"
                    binding.txtPost.text = String.format(Locale.getDefault(), "%d  %s  • ", profile!!.e0a_NOG, tips)
                    binding.txtWon.text = String.format(Locale.getDefault(), "%d  won  • ", profile!!.e0b_WG)
                    binding.txtAccuracy.text = String.format(Locale.getDefault(), "%.1f%%", profile!!.e0c_WGP.toDouble())

                    //set Display picture
                    Glide.with(applicationContext).load(profile!!.b2_dpUrl)
                            .placeholder(R.drawable.dummy)
                            .error(getPlaceholderImage(userId!![0]))
                            .into(binding.imgDp)
                    if (!performanceList.isEmpty()) return@addOnSuccessListener
                    if (profile!!.e0a_NOG > 0) {
                        for (i in 1..6) {
                            val row = getRow(i)
                            if (!row.isEmpty()) performanceList.add(row)
                        }
                        binding.performanceList.adapter = PerformanceAdapter(performanceList)
                    }
                }
    }

    override fun onClick(v: View) {
        val intent = Intent(this, FollowerListActivity::class.java)
        intent.putExtra("personId", userId)
        when (v.id) {
            R.id.imgDp -> showDp()
            R.id.lnrFollowers -> {
                if (Integer.valueOf(binding.txtFollowers.text.toString()) < 1) return
                intent.putExtra("search_type", "followers")
                startActivity(intent)
            }
            R.id.lnrFollowing -> {
                if (Integer.valueOf(binding.txtFollowing.text.toString()) < 1) return
                intent.putExtra("search_type", "followings")
                startActivity(intent)
            }
            R.id.lnrSubscribers -> {
                if (Integer.valueOf(binding.txtSubscribers.text.toString()) < 1) return
                intent.putExtra("search_type", "subscribers")
                startActivity(intent)
            }
            R.id.lnrSubscription -> {
                if (Integer.valueOf(binding.txtSubscription.text.toString()) < 1) return
                intent.putExtra("search_type", "subscribed_to")
                startActivity(intent)
            }
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        finish()
        return true
    }

}